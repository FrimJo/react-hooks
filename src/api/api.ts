let DB_USER = { id: 5, name: 'Stina' }

export type User = { id: number; name: string }

export const api = {
  authenticateAsync: (variables: { username: string; password: string }) => {
    console.log('authenticateAsync POST')
    return new Promise<User>((resolve, reject) => {
      setTimeout(() => {
        console.log('logInAsync resolve void', variables)
        DB_USER = { id: 5, name: variables.username }
        resolve(DB_USER)
      }, 2000)
    })
  },
  updateUserAsync: (user: User) => {
    console.log('updateUserAsync POST')
    return new Promise<User>((resolve, reject) => {
      setTimeout(() => {
        DB_USER = user
        console.log('updateUserAsync resolve', DB_USER)
        resolve(DB_USER)
        // console.log('updateUserAsync reject', user)
        // reject('Could not fetch user')
      }, 2000)
    })
  },
  getUserAsync: (id: number) => {
    console.log('getUserAsync POST')
    return new Promise<User>((resolve, reject) => {
      setTimeout(() => {
        console.log('getUserAsync resolve', { ...DB_USER, id })
        resolve({ ...DB_USER, id })
      }, 2000)
    })
  },
}
