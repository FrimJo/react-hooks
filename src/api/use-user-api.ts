/* eslint-disable @typescript-eslint/camelcase */
import { useHttpMiddleware } from 'utilities/use-http-middleware'
import React from 'react'

export type User = {
  id: number
  email: string
  firstName: string
  lastName: string
  avatar?: string
}
export type ApiUser = {
  id: number
  email: string
  first_name: string
  last_name: string
  avatar?: string
}

const API_BASE_URL = 'https://reqres.in'

export function mapApiUserToClientUser(apiUser: ApiUser): User {
  return {
    id: apiUser.id,
    email: apiUser.email,
    firstName: apiUser.first_name,
    lastName: apiUser.last_name,
    avatar: apiUser.avatar,
  }
}

function mapClientUserToApiUser(clientUser: Partial<User>): Partial<ApiUser> {
  return {
    id: clientUser.id,
    email: clientUser.email,
    first_name: clientUser.firstName,
    last_name: clientUser.lastName,
    avatar: clientUser.avatar,
  }
}

export const useUserApi = () => {
  const initOptions: RequestInit = React.useMemo(() => ({}), [])
  const http = useHttpMiddleware(initOptions)

  const update = React.useCallback(
    (user: Partial<User>): Promise<Partial<User> & { updatedAt: Date }> => {
      console.log('update', user)
      const content_ = JSON.stringify(mapClientUserToApiUser(user))

      const init: RequestInit = {
        body: content_,
        method: 'PATCH',
      }
      return http
        .fetch(API_BASE_URL + `/api/users/${user.id}`, init)
        .then((response) => response.json())
        .then((result) => result)
    },
    [http]
  )

  const getSingle = React.useCallback(
    (id: number): Promise<User> => {
      const init: RequestInit = {
        method: 'GET',
      }
      return http
        .fetch(API_BASE_URL + `/api/users/${id}`, init)
        .then((response) => response.json())
        .then((result) => result.data)
        .then(mapApiUserToClientUser)
    },
    [http]
  )

  return React.useMemo(
    () => ({
      update,
      getSingle,
    }),
    [getSingle, update]
  )
}
