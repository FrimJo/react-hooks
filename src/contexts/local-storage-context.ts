import * as React from 'react'
import { asContext } from 'utilities/as-context'

type LocalStorgateType = { [key in string]: any }
const LOCAL_STORAGE_KEY = 'example-app'

export function getLocalStorage() {
  const itemJSON = window.localStorage.getItem(LOCAL_STORAGE_KEY)
  return itemJSON == null ? {} : JSON.parse(itemJSON)
}

export function setLocalStorage(
  value: ((prevValue: LocalStorgateType) => LocalStorgateType) | LocalStorgateType
): void {
  const prevValues = getLocalStorage()
  // Allow value to be a function so we have same API as useState
  const valueToStore = value instanceof Function ? value(prevValues) : value
  // Save to local storage
  return window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(valueToStore))
}

function useLocalStorageContext() {
  const [storedValues, setStoredValues] = React.useState<LocalStorgateType>(() => getLocalStorage())

  const setValue = React.useCallback(
    function <ValueType>(key: string, value: ((prevValue: ValueType) => ValueType) | ValueType) {
      // Allow value to be a function so we have same API as useState
      const valueToStore = value instanceof Function ? value(storedValues[key]) : value

      // Save state
      setStoredValues((prevValues) => {
        const updatedValues = { ...prevValues, [key]: valueToStore }
        // Save to local storage
        window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(updatedValues))
        return updatedValues
      })
    },
    [storedValues]
  )

  const clearValue = React.useCallback((key: string) => {
    setStoredValues((prevValues: any) => {
      const values = { ...prevValues }
      delete values[key]
      window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(values))
      return values
    })
  }, [])

  return React.useMemo(() => ({ state: { storedValues }, actions: { setValue, clearValue } }), [
    clearValue,
    setValue,
    storedValues,
  ])
}

const LocalStorageContext = asContext(useLocalStorageContext, 'LocalStorageContext')
export const LocalStorageProvider = LocalStorageContext.Provider
export const useLocalStorageActions = LocalStorageContext.useActions
export const useLocalStorageState = LocalStorageContext.useState
