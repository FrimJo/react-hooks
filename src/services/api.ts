import axios from 'axios'

export type Payload = Readonly<{
  type: string
  length: number
  data: number[]
  success: boolean
}>

export const randomNumberUrl =
  'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8'

export const fetchRandomNumberAsync = async (): Promise<Payload> =>
  await axios
    .get<Payload>('https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8')
    .then(response => response.data)
// await fetch('https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8', { signal }).then(response => response.json())
