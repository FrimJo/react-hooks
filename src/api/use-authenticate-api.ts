import React from 'react'
import { useHttpMiddleware } from 'utilities/use-http-middleware'
import { User, mapApiUserToClientUser } from './use-user-api'

const API_BASE_URL = 'https://reqres.in'

export const useAuthenticateApi = () => {
  const initOptions: RequestInit = React.useMemo(() => ({}), [])
  const http = useHttpMiddleware(initOptions)

  const register = React.useCallback(
    (credentials: { email: string; password: string }): Promise<{ id: number; token: string }> => {
      const content_ = JSON.stringify(credentials)

      const init: RequestInit = {
        body: content_,
        method: 'POST',
      }

      return http.fetch(API_BASE_URL + '/api/register', init).then(async (response) => {
        const json = response.json()
        if (response.status === 400) {
          throw await json
        }
        return json
      })
    },
    [http]
  )

  const login = React.useCallback(
    (credentials: { email: string; password: string }): Promise<{ token: string }> => {
      const content_ = JSON.stringify(credentials)

      const init: RequestInit = {
        body: content_,
        method: 'POST',
      }

      return http.fetch(API_BASE_URL + '/api/login', init).then(async (response) => {
        const json = response.json()
        if (response.status === 400) {
          throw await json
        }
        return json
      })
    },
    [http]
  )

  const getLoggedInUser = React.useCallback(
    (id: number): Promise<User> => {
      const init: RequestInit = {
        method: 'GET',
      }
      return http
        .fetch(API_BASE_URL + `/api/users/${id}`, init)
        .then((response) => response.json())
        .then((result) => result.data)
        .then(mapApiUserToClientUser)
    },
    [http]
  )

  return React.useMemo(
    () => ({
      login,
      register,
      getLoggedInUser,
    }),
    [login, register, getLoggedInUser]
  )
}
