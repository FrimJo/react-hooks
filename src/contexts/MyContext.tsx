import useAxiosSuspense from 'hooks/useAxiosSuspense'
import React, { useEffect, useMemo } from 'react'
import useNumber from '../hooks/useNumber'
import * as api from '../services/api'

interface IState {
  readonly value: number
}
type Actions = Readonly<{
  increment: () => void
  reset: () => void
  setValue: (value: number) => void
}>

const StateContext = React.createContext({} as IState)
const ActionsContext = React.createContext({} as Actions)

const MyContextProvider: React.FunctionComponent = ({ children }) => {
  const { value, increment, reset, setValue } = useNumber(
    (currentState, action) => action.changes
  )

  const [data] = useAxiosSuspense<api.Payload>(
    'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint8'
  )
  console.log('data', data)

  useEffect(() => {
    if (data.data.length > 0) {
      setValue(data.data[0])
    }
  }, [data, setValue])

  const stateValue: IState = useMemo(
    () => ({
      value,
    }),
    [value]
  )

  const actionValue: Actions = useMemo(
    () => ({
      increment,
      reset,
      setValue,
    }),
    [increment, reset, setValue]
  )
  return (
    <StateContext.Provider value={stateValue}>
      <ActionsContext.Provider value={actionValue}>
        {children}
      </ActionsContext.Provider>
    </StateContext.Provider>
  )
}

export { StateContext, ActionsContext, MyContextProvider }
