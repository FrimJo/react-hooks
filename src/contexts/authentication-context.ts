import React from 'react'
import { queryCache, useMutation } from 'react-query'
import { buildContext } from 'utilities/build-context'
import { useAuthenticateApi } from 'api/use-authenticate-api'
import { useTokenData } from 'utilities/token-data'
import { useRegisterStore } from 'utilities/use-register-store'

const useAuthentication = () => {
  const token = useTokenData()
  const registerStore = useRegisterStore()
  const api = useAuthenticateApi()

  const [loginFromApi, loginState] = useMutation(api.login, { throwOnError: true })
  const [registerFromApi, registerState] = useMutation(api.register, { throwOnError: true })

  const isLoggedIn = React.useMemo(() => !!token.storage, [token.storage])

  const logout = React.useCallback(() => {
    queryCache.clear()
    token.clear()
  }, [token])

  const login = React.useCallback(
    (credentials: { username: string; password: string }) =>
      loginFromApi({
        email: credentials.username,
        password: credentials.password,
      }).then((tokenData) => {
        console.log('registerStore.storage', registerStore.storage)
        if (!registerStore.storage) {
          return Promise.reject('Need to register first.')
        }
        const { id: userId } = registerStore.storage
        return queryCache
          .prefetchQuery(['user', userId], (key, id) => api.getLoggedInUser(id))
          .then((user) => {
            token.set({ ...tokenData, id: userId })
            return user
          })
      }),
    [api, loginFromApi, registerStore.storage, token]
  )

  const register = React.useCallback(
    (credentials: { email: string; password: string }) =>
      registerFromApi(credentials).then(({ id }) => registerStore.set({ id })),
    [registerFromApi, registerStore]
  )

  return {
    state: { isLoggedIn, login: loginState, register: registerState },
    actions: { login, logout, register },
  }
}

const Context = buildContext(useAuthentication, 'AuthenticationContext')

const useAuthenticatedState = () => {
  return React.useContext(Context.StateContext)
  // const state = React.useContext(Context.StateContext)
  // if (state === undefined) {
  //   throw Error('Missing AuthenticationProvider for useAuthenticationState')
  // }

  // if (!state.isLoggedIn) {
  //   throw Error('User not logged in')
  // }

  // if (!state.loggedInUser) {
  //   throw Error('Missing user id in AuthenticationContext.provider')
  // }

  // return {
  //   ...state,
  //   loggedInUser: state.loggedInUser,
  // }
}

export const AuthenticationContext = { ...Context, useAuthenticatedState }
