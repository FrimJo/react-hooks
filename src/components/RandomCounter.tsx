import React, { useCallback, useContext } from 'react'
import styled from 'styled-components'
import { ActionsContext, StateContext } from '../contexts/MyContext'

const StyledButton = styled.button``

const RandomCounter: React.FunctionComponent = ({ children }) => {
  const context = useContext(StateContext)
  const actions = useContext(ActionsContext)

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) =>
      actions.setValue(+event.target.value),
    [actions]
  )

  return (
    <>
      <div>{`Value: ${context.value}`}</div>
      <StyledButton onClick={actions.increment}>Increment</StyledButton>
      <StyledButton onClick={actions.reset}>Clear</StyledButton>
      <input
        aria-label="input-field"
        value={context.value}
        onChange={handleChange}
      />
    </>
  )
}

export default RandomCounter
