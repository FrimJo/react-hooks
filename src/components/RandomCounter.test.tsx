import { AxiosStaticMock } from '__mocks__/axios'
import axios from 'axios'
import 'jest-dom/extend-expect'
import React, { Suspense } from 'react'
import { act, cleanup, fireEvent, render, wait } from 'react-testing-library'
import { Payload } from 'services/api'
import { MyContextProvider } from '../contexts/MyContext'
import RandomCounter from './RandomCounter'

afterEach(cleanup)

const axiosMock = axios as AxiosStaticMock

test('RandomCounter makes an API call and displays the value', async () => {
  const data: Payload = {
    type: 'number',
    length: 2,
    data: [5],
    success: true,
  }

  axiosMock.get.mockResolvedValueOnce({ data })

  const { getByText, getByLabelText, container } = render(
    <Suspense fallback={<>loadingâ€¦</>}>
      <MyContextProvider>
        <RandomCounter />
      </MyContextProvider>
    </Suspense>
  )

  await wait()

  const inputField = getByLabelText('input-field') as HTMLInputElement
  const incrementButton = getByText(/Increment/i) as HTMLButtonElement
  const clearButton = getByText(/Clear/i) as HTMLButtonElement
  const valueText = getByText(/Value/i) as HTMLButtonElement

  expect(axiosMock.get).toHaveBeenCalledTimes(1)
  expect(valueText).toHaveTextContent('5')
  expect(inputField.value).toBe('5')
  expect(container.firstChild).toMatchSnapshot('idle')

  act(() => {
    fireEvent.change(inputField, { target: { value: '34' } })
  })

  expect(inputField.value).toBe('34')
  act(() => {
    fireEvent.click(incrementButton)
  })
  expect(inputField.value).toBe('35')

  act(() => {
    fireEvent.click(clearButton)
  })
  expect(inputField.value).toBe('0')
})
