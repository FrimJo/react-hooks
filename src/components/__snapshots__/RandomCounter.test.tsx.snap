// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomCounter makes an API call and displays the value: idle 1`] = `
<div>
  Value: 5
</div>
`;
